File: sdn-server/internal/channels/channels_test.go
package channels

import "testing"

var testInternalSchemaSuffix = string([]byte{'.', 'f', 'b', 's'})

func TestParseChannelIDParsesFromRight(t *testing.T) {
	t.Parallel()

	tests := []struct {
		input        string
		sourceID     string
		standardCode string
		feedUUID     string
	}{
		{
			input:        "celestrak-OMM",
			sourceID:     "celestrak",
			standardCode: "OMM",
		},
		{
			input:        "celestrak-eth-CDM",
			sourceID:     "celestrak-eth",
			standardCode: "CDM",
		},
		{
			input:        "spaceaware-live-OMM-550e8400-e29b-41d4-a716-446655440000",
			sourceID:     "spaceaware-live",
			standardCode: "OMM",
			feedUUID:     "550e8400-e29b-41d4-a716-446655440000",
		},
	}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.input, func(t *testing.T) {
			t.Parallel()
			got, err := ParseChannelID(tc.input)
			if err != nil {
				t.Fatalf("ParseChannelID(%q) returned error: %v", tc.input, err)
			}
			if got.ChannelID != tc.input || got.SourceID != tc.sourceID || got.StandardCode != tc.standardCode || got.FeedUUID != tc.feedUUID {
				t.Fatalf("ParseChannelID(%q) = %+v", tc.input, got)
			}
		})
	}
}

func TestParseChannelIDRejectsInvalidPublicNames(t *testing.T) {
	t.Parallel()

	for _, input := range []string{
		"",
		"OMM",
		"-OMM",
		"celestrak-",
		"celestrak-omm",
		"celestrak-OMM" + testInternalSchemaSuffix,
		"celestrak-OMM-not-a-uuid",
	} {
		input := input
		t.Run(input, func(t *testing.T) {
			t.Parallel()
			if _, err := ParseChannelID(input); err == nil {
				t.Fatalf("ParseChannelID(%q) succeeded", input)
			}
		})
	}
}

func TestChannelFormattingAndInternalSchemaMapping(t *testing.T) {
	t.Parallel()

	got, err := FormatChannelID(ChannelIDInput{SourceID: "celestrak-eth", StandardCode: "CDM"})
	if err != nil {
		t.Fatalf("FormatChannelID returned error: %v", err)
	}
	if got != "celestrak-eth-CDM" {
		t.Fatalf("FormatChannelID = %q", got)
	}

	got, err = FormatChannelID(ChannelIDInput{
		SourceID:     "spaceaware",
		StandardCode: "OMM",
		FeedUUID:     "550e8400-e29b-41d4-a716-446655440000",
	})
	if err != nil {
		t.Fatalf("FormatChannelID with UUID returned error: %v", err)
	}
	if got != "spaceaware-OMM-550e8400-e29b-41d4-a716-446655440000" {
		t.Fatalf("FormatChannelID with UUID = %q", got)
	}

	schemaName, err := SchemaNameFromStandardCode("OMM")
	if err != nil {
		t.Fatalf("SchemaNameFromStandardCode returned error: %v", err)
	}
	if schemaName != "OMM"+testInternalSchemaSuffix {
		t.Fatalf("SchemaNameFromStandardCode = %q", schemaName)
	}

	standardCode, err := StandardCodeFromSchemaName("OMM" + testInternalSchemaSuffix)
	if err != nil {
		t.Fatalf("StandardCodeFromSchemaName returned error: %v", err)
	}
	if standardCode != "OMM" {
		t.Fatalf("StandardCodeFromSchemaName = %q", standardCode)
	}
	if DiscoveryTopic("OMM") != "/spacedatanetwork/channels/OMM" {
		t.Fatalf("DiscoveryTopic returned unexpected topic")
	}
}

File: sdn-js/src/channels.test.ts
import { describe, expect, it } from 'vitest';
import {
  channelDiscoveryTopic,
  formatChannelId,
  parseChannelId,
  schemaNameFromStandardCode,
  standardCodeFromSchemaName,
} from './channels';

const internalSchemaSuffix = String.fromCharCode(46, 102, 98, 115);

describe('SDN channel naming', () => {
  it('parses source and three-letter standard code from the right', () => {
    expect(parseChannelId('celestrak-OMM')).toEqual({
      channelId: 'celestrak-OMM',
      sourceId: 'celestrak',
      standardCode: 'OMM',
      feedUuid: null,
    });
    expect(parseChannelId('celestrak-eth-CDM')).toEqual({
      channelId: 'celestrak-eth-CDM',
      sourceId: 'celestrak-eth',
      standardCode: 'CDM',
      feedUuid: null,
    });
  });

  it('parses optional UUID suffix without splitting hyphenated source IDs', () => {
    expect(parseChannelId('spaceaware-live-OMM-550e8400-e29b-41d4-a716-446655440000')).toEqual({
      channelId: 'spaceaware-live-OMM-550e8400-e29b-41d4-a716-446655440000',
      sourceId: 'spaceaware-live',
      standardCode: 'OMM',
      feedUuid: '550e8400-e29b-41d4-a716-446655440000',
    });
  });

  it('rejects lowercase, missing pieces, and schema-file suffixes', () => {
    for (const input of [
      '',
      'OMM',
      '-OMM',
      'celestrak-',
      'celestrak-omm',
      `celestrak-OMM${internalSchemaSuffix}`,
      'celestrak-OMM-not-a-uuid',
    ]) {
      expect(() => parseChannelId(input), input).toThrow();
    }
  });

  it('formats channels and maps standard codes to internal schema names', () => {
    expect(formatChannelId({ sourceId: 'celestrak-eth', standardCode: 'CDM' })).toBe('celestrak-eth-CDM');
    expect(formatChannelId({
      sourceId: 'spaceaware',
      standardCode: 'OMM',
      feedUuid: '550e8400-e29b-41d4-a716-446655440000',
    })).toBe('spaceaware-OMM-550e8400-e29b-41d4-a716-446655440000');
    expect(schemaNameFromStandardCode('OMM')).toBe(`OMM${internalSchemaSuffix}`);
    expect(standardCodeFromSchemaName(`OMM${internalSchemaSuffix}`)).toBe('OMM');
    expect(channelDiscoveryTopic('OMM')).toBe('/spacedatanetwork/channels/OMM');
  });
});
